dusk-labyrinth.lic

```diff
@@ -7,74 +7,208 @@ custom_require.call %w[common common-items events]
+    message('Try to move yourself to another room then restart script')
+    exit
   end
 
+  # Moves to a nearby room in the maze.
+  # Prioritizes rooms we have not been in yet.
+  # Skips over rooms we've been in before if looking for a pet.
   def wander
     waitrt?
-    next_dir = next_move(XMLData.room_exits)
-    @dir_to_prev_dir.each_key { |dir| break if move(dir) } unless move(next_dir)
-    @prev_dir = next_dir
+
+    @current_room = create_room if @current_room.nil?
+    echo("*** [wander] @current_room: #{@current_room}") if $debug_mode_dl
+
+    @forward_move = choose_next_move(@current_room['exits'])
+    @reverse_move = @reverse_direction_map[@forward_move]
+
+    echo("*** [wander] forward_move: #{@forward_move}") if $debug_mode_dl
+    echo("*** [wander] reverse_move: #{@reverse_move}") if $debug_mode_dl
+
+    move(@forward_move)
+
+    # Since rooms in the maze don't have ids, we build
+    # a bi-directional graph where each vertex (room)
+    # points to another vertex (room) and vice versa.
+    @previous_room = @current_room
+    @current_room = @current_room['exits'][@forward_move]['room'] || create_room
+
+    @previous_room['exits'][@forward_move]['room'] = @current_room
+    @current_room['exits'][@reverse_move]['room'] = @previous_room
+
+    # Increment the count of us having moved between these two rooms.
+    # In graph algorithms, this is known as a weighted graph so that
+    # when we pick the "next best move" to go down we choose paths
+    # with the least count (weight) thereby prioritizing paths we
+    # have not visited before spending time visiting rooms we've been to.
+    @previous_room['exits'][@forward_move]['count'] = @previous_room['exits'][@forward_move]['count'] + 1
+    @current_room['exits'][@reverse_move]['count'] = @current_room['exits'][@reverse_move]['count'] + 1
+
   end
 
+  def create_room
+    new_room = {
+      'exits' => {}
+    }
+    # Initialize weights of each path out of this room.
+    checkpaths.each do |dir|
+      new_room['exits'][dir] = {
+        # How many times we've traversed this path.
+        'count' => 0,
+        # Which room this path moves you to.
+        # If not null then this room has an exit
+        # in the reverse direction that points back
+        # to this room we are creating.
+        # In this manner two rooms have bi-directional
+        # edges to the other room.
+        # This is critical to the maze mapping algorithm
+        # because each room has no roomid in the labyrinth
+        # so this is how we keep track in memory of which
+        # rooms we have visited.
+        'room' => nil
+      }
+    end
+    return new_room
+  end
+
+  # Returns a direction to move in.
+  # Prioritizes rooms we've not visited before.
+  def choose_next_move(exits)
+    # https://medium.com/@florenceliang/some-notes-about-using-hash-sort-by-in-ruby-f4b3a700fc33
+    sorted_exits = exits.sort_by { |k, v| v['count'] }
```

Comment:
Each "exit" in the array given is a "room" (vertex) in our graph. And the rooms (vertices) keep track of how many times we've moved from the the previous room to this room in the `count` property. So of the exits from this room, we want to pick a direction that we've not been down before, or if we've visited them all, begin backtracking rooms that have been visited the least.